src/PrestaShopBundle/Routing/LegacyUrlConverter.php

```diff
@@ -0,0 +1,268 @@
+<?php
+/**
+ * 2007-2018 PrestaShop.
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2018 PrestaShop SA
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShopBundle\Routing;
+
+use PrestaShopBundle\Routing\Exception\ArgumentException;
+use PrestaShopBundle\Routing\Exception\RouteNotFoundException;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+use Symfony\Component\Routing\Route;
+use Symfony\Component\Routing\RouterInterface;
+
+/**
+ * Class LegacyUrlConverter is able to convert query parameters or an url into a
+ * migrated Symfony url. It uses information set in the routes via _legacy_link
+ * to do so.
+ */
+final class LegacyUrlConverter
+{
+    /**
+     * @var RouterInterface
+     */
+    private $router;
+
+    /**
+     * @var array
+     */
+    private $legacyRoutes;
+
+    /**
+     * @var array
+     */
+    private $controllersActions;
+
+    /**
+     * LegacyUrlConverter constructor.
+     *
+     * @param RouterInterface $router
+     */
+    public function __construct(RouterInterface $router)
+    {
+        $this->router = $router;
+        $this->legacyRoutes = [];
+        $this->controllersActions = [];
+        /** @var Route $route */
+        foreach ($this->router->getRouteCollection() as $routeName => $route) {
+            $this->addRoute($routeName, $route);
+        }
```

Comment:
How long does it take to run this loop?

If the time needed is noticeable. It could be better to run it at the first real use of this class.

Reply:
I don't think it is that long since it only creates a few arrays, but I can delay this creation via getters sure, it's probable for the best
Although since this service is called pretty early in the Listener it will have to perform it anyways, so I think the gain will be small